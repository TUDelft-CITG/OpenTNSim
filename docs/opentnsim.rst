=========
OpenTNSim
=========

This page lists all functions and classes available in the OpenTNSim.model and OpenTNSim.core modules. For examples on how to use these submodules please check out the Examples page, information on installing OpenTNim can be found on the Installation page.

Submodules
----------

The main components are the Model module and the Core module. All of their components are listed below. 

opentnsim\.model module
--------------------------

.. automodule:: opentnsim.model
   :members:
   :undoc-members:
   :show-inheritance:

opentnsim\.core module
--------------------------

.. automodule:: opentnsim.core
   :members:
   :undoc-members:
   :show-inheritance:

Module contents
---------------

.. automodule:: opentnsim
   :members:
   :undoc-members:
   :show-inheritance:
