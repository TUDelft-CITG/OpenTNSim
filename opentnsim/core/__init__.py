from .resource import HasResource
from .misc import Neighbours, ExtraMetadata
from .vessel_properties import HasLength, HasLoad, VesselProperties
from .container import HasContainer
from .movable import Movable, ContainerDependentMovable, Routable, Routeable
