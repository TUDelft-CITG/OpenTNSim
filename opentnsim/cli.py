# -*- coding: utf-8 -*-

"""Console script for opentnsim."""
# package(s) related to the command line interface
import sys

import click

import opentnsim.server


@click.group()
def cli(args=None):
    """OpenTNSim simulation."""
    click.echo("Replace this message by putting your code into " "opentnsim.cli.main")
    click.echo("See click documentation at http://click.pocoo.org/")
    return 0


@cli.command()
@click.option("--host", default="0.0.0.0")
@click.option("--port", default=5000, type=int)
@click.option("--debug/--no-debug", default=False)
def serve(host, port, debug, args=None):
    """Run a flask server with the backend code"""
    app = opentnsim.server.app
    app.run(host=host, debug=debug, port=port)


if __name__ == "__main__":
    sys.exit(cli())  # pragma: no cover
