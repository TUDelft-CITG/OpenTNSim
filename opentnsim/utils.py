"""
Utilities for OpenTNSim

This file also includes the networkx read shapefile functions that used to be in networkx.
These originate from:
https://github.com/networkx/networkx/blob/9256ef670730b741105a9264197353004bd6719f/networkx/readwrite/nx_shp.py

Generates a networkx.DiGraph from point and line shapefiles.

"The Esri Shapefile or simply a shapefile is a popular geospatial vector
data format for geographic information systems software. It is developed
and regulated by Esri as a (mostly) open specification for data
interoperability among Esri and other software products."
See https://en.wikipedia.org/wiki/Shapefile for additional information.
"""

import warnings

import networkx as nx
import shapely

import opentnsim


class NetworkWarning(Warning):
    pass


def find_notebook_path():
    """Lookup the path where the notebooks are located. Returns a pathlib.Path object."""
    opentnsim_path = pathlib.Path(opentnsim.__file__)
    # check if the path looks normal
    assert "opentnsim" in str(opentnsim_path), "we can't find the opentnsim path: {opentnsim_path} (opentnsim not in path name)"
    # src_dir/opentnsim/__init__.py -> ../.. -> src_dir
    src_path = opentnsim_path.parent.parent
    notebook_path = opentnsim_path.parent.parent / "notebooks"
    return notebook_path


def network_check(graph):
    """Assertions about the graphs used in OpenTNSim"""
    node_type = (str, shapely.Point)
    ok = True

    if not isinstance(graph, nx.Graph):
        warnings.warn("graph should be of type nx.Graph", NetworkWarning)
        ok = False
    if len(graph.nodes) < 2:
        warnings.warn("there should be at least 2 nodes in the graph", NetworkWarning)
        ok = False
    if len(graph.edges) < 1:
        warnings.warn("there should be at least 1 edge in the graph", NetworkWarning)
        ok = False
    if not all(isinstance(n, node_type) for n in graph.nodes.keys()):
        warnings.warn("all keys should be str or Point", NetworkWarning)
        ok = False
    # check all edges
    for e in graph.edges.keys():
        if len(e) != 2:
            warnings.warn(f"edge keys should be tuples of length 2, {e} was not", NetworkWarning)
            ok = False
            # stop checking
            break

        source, _ = e
        if not isinstance(source, node_type):
            warnings.warn(f"edges should be of a tuple of Points or str, {e} was not ", NetworkWarning)
            ok = False
            # stop checking
            break
    for e, edge in graph.edges.items():
        if "geometry" not in edge:
            warnings.warn(f"edges should have of geometry attribute, {e} did not ", NetworkWarning)
            ok = False
        elif not isinstance(edge["geometry"], (str, shapely.Geometry)):
            warnings.warn(f"edges geometry should be of type string or Geometry, {edge['geometry']} was not.", NetworkWarning)
            ok = False
    return ok


# Ignore functions copied from networkx
# // START-NOSCAN
def read_shp(path, simplify=True, geom_attrs=True, strict=True):
    """Generates a networkx.DiGraph from shapefiles.

       read_shp used to be part of NetworkX.
       See https://networkx.org/documentation/latest/auto_examples/index.html#geospatial.

    Point geometries are
    translated into nodes, lines into edges. Coordinate tuples are used as
    keys. Attributes are preserved, line geometries are simplified into start
    and end coordinates. Accepts a single shapefile or directory of many
    shapefiles.

    "The Esri Shapefile or simply a shapefile is a popular geospatial vector
    data format for geographic information systems software [1]_."

    Parameters
    ----------
    path : file or string
       File, directory, or filename to read.

    simplify:  bool
        If True, simplify line geometries to start and end coordinates.
        If False, and line feature geometry has multiple segments, the
        non-geometric attributes for that feature will be repeated for each
        edge comprising that feature.

    geom_attrs: bool
        If True, include the Wkb, Wkt and Json geometry attributes with
        each edge.

        NOTE:  if these attributes are available, write_shp will use them
        to write the geometry.  If nodes store the underlying coordinates for
        the edge geometry as well (as they do when they are read via
        this method) and they change, your geomety will be out of sync.

    strict: bool
        If True, raise NetworkXError when feature geometry is missing or
        GeometryType is not supported.
        If False, silently ignore missing or unsupported geometry in features.

    Returns
    -------
    G : NetworkX graph

    Raises
    ------
    ImportError
       If ogr module is not available.

    RuntimeError
       If file cannot be open or read.

    NetworkXError
       If strict=True and feature is missing geometry or GeometryType is
       not supported.

    Examples
    --------
    >>> G = nx.read_shp("test.shp")  # doctest: +SKIP

    References
    ----------
    .. [1] https://en.wikipedia.org/wiki/Shapefile
    """
    try:
        from osgeo import ogr
    except ImportError as err:
        raise ImportError("read_shp requires OGR: https://www.gdal.org/") from err

    if not isinstance(path, str):
        return

    net = nx.DiGraph()
    shp = ogr.Open(path)
    if shp is None:
        raise RuntimeError(f"Unable to open {path}")
    for lyr in shp:
        fields = [x.GetName() for x in lyr.schema]
        for f in lyr:
            g = f.geometry()
            if g is None:
                if strict:
                    raise nx.NetworkXError("Bad data: feature missing geometry")
                else:
                    continue
            flddata = [f.GetField(f.GetFieldIndex(x)) for x in fields]
            attributes = dict(zip(fields, flddata))
            attributes["ShpName"] = lyr.GetName()
            # Note:  Using layer level geometry type
            if g.GetGeometryType() == ogr.wkbPoint:
                net.add_node((g.GetPoint_2D(0)), **attributes)
            elif g.GetGeometryType() in (ogr.wkbLineString, ogr.wkbMultiLineString):
                for edge in edges_from_line(g, attributes, simplify, geom_attrs):
                    e1, e2, attr = edge
                    net.add_edge(e1, e2)
                    net[e1][e2].update(attr)
            else:
                if strict:
                    raise nx.NetworkXError(f"GeometryType {g.GetGeometryType()} not supported")

    return net


def edges_from_line(geom, attrs, simplify=True, geom_attrs=True):
    """
    Generate edges for each line in geom
    Written as a helper for read_shp

    Parameters
    ----------

    geom:  ogr line geometry
        To be converted into an edge or edges

    attrs:  dict
        Attributes to be associated with all geoms

    simplify:  bool
        If True, simplify the line as in read_shp

    geom_attrs:  bool
        If True, add geom attributes to edge as in read_shp


    Returns
    -------
     edges:  generator of edges
        each edge is a tuple of form
        (node1_coord, node2_coord, attribute_dict)
        suitable for expanding into a networkx Graph add_edge call

    .. deprecated:: 2.6
    """
    msg = (
        "edges_from_line is deprecated and will be removed in 3.0."
        "See https://networkx.org/documentation/latest/auto_examples/index.html#geospatial."
    )
    warnings.warn(msg, DeprecationWarning, stacklevel=2)
    try:
        from osgeo import ogr
    except ImportError as err:
        raise ImportError("edges_from_line requires OGR: " "https://www.gdal.org/") from err

    if geom.GetGeometryType() == ogr.wkbLineString:
        if simplify:
            edge_attrs = attrs.copy()
            last = geom.GetPointCount() - 1
            if geom_attrs:
                edge_attrs["Wkb"] = geom.ExportToWkb()
                edge_attrs["Wkt"] = geom.ExportToWkt()
                edge_attrs["Json"] = geom.ExportToJson()
            yield (geom.GetPoint_2D(0), geom.GetPoint_2D(last), edge_attrs)
        else:
            for i in range(0, geom.GetPointCount() - 1):
                pt1 = geom.GetPoint_2D(i)
                pt2 = geom.GetPoint_2D(i + 1)
                edge_attrs = attrs.copy()
                if geom_attrs:
                    segment = ogr.Geometry(ogr.wkbLineString)
                    segment.AddPoint_2D(pt1[0], pt1[1])
                    segment.AddPoint_2D(pt2[0], pt2[1])
                    edge_attrs["Wkb"] = segment.ExportToWkb()
                    edge_attrs["Wkt"] = segment.ExportToWkt()
                    edge_attrs["Json"] = segment.ExportToJson()
                    del segment
                yield (pt1, pt2, edge_attrs)

    elif geom.GetGeometryType() == ogr.wkbMultiLineString:
        for i in range(geom.GetGeometryCount()):
            geom_i = geom.GetGeometryRef(i)
            yield from edges_from_line(geom_i, attrs, simplify, geom_attrs)


def write_shp(G, outdir):
    """Writes a networkx.DiGraph to two shapefiles, edges and nodes.

       write_shp used to be part of networx.
       See https://networkx.org/documentation/latest/auto_examples/index.html#geospatial.

    Nodes and edges are expected to have a Well Known Binary (Wkb) or
    Well Known Text (Wkt) key in order to generate geometries. Also
    acceptable are nodes with a numeric tuple key (x,y).

    "The Esri Shapefile or simply a shapefile is a popular geospatial vector
    data format for geographic information systems software [1]_."

    Parameters
    ----------
    G : NetworkX graph
        Directed graph
    outdir : directory path
       Output directory for the two shapefiles.

    Returns
    -------
    None

    Examples
    --------
    nx.write_shp(digraph, '/shapefiles') # doctest +SKIP

    References
    ----------
    .. [1] https://en.wikipedia.org/wiki/Shapefile
    """
    try:
        from osgeo import ogr
    except ImportError as err:
        raise ImportError("write_shp requires OGR: https://www.gdal.org/") from err
    # easier to debug in python if ogr throws exceptions
    ogr.UseExceptions()

    def netgeometry(key, data):
        if "Wkb" in data:
            geom = ogr.CreateGeometryFromWkb(data["Wkb"])
        elif "Wkt" in data:
            geom = ogr.CreateGeometryFromWkt(data["Wkt"])
        elif type(key[0]).__name__ == "tuple":  # edge keys are packed tuples
            geom = ogr.Geometry(ogr.wkbLineString)
            _from, _to = key[0], key[1]
            try:
                geom.SetPoint(0, *_from)
                geom.SetPoint(1, *_to)
            except TypeError:
                # assume user used tuple of int and choked ogr
                _ffrom = [float(x) for x in _from]
                _fto = [float(x) for x in _to]
                geom.SetPoint(0, *_ffrom)
                geom.SetPoint(1, *_fto)
        else:
            geom = ogr.Geometry(ogr.wkbPoint)
            try:
                geom.SetPoint(0, *key)
            except TypeError:
                # assume user used tuple of int and choked ogr
                fkey = [float(x) for x in key]
                geom.SetPoint(0, *fkey)

        return geom

    # Create_feature with new optional attributes arg (should be dict type)
    def create_feature(geometry, lyr, attributes=None):
        feature = ogr.Feature(lyr.GetLayerDefn())
        feature.SetGeometry(g)
        if attributes is not None:
            # Loop through attributes, assigning data to each field
            for field, data in attributes.items():
                feature.SetField(field, data)
        lyr.CreateFeature(feature)
        feature.Destroy()

    # Conversion dict between python and ogr types
    OGRTypes = {int: ogr.OFTInteger, str: ogr.OFTString, float: ogr.OFTReal}

    # Check/add fields from attribute data to Shapefile layers
    def add_fields_to_layer(key, value, fields, layer):
        # Field not in previous edges so add to dict
        if type(value) in OGRTypes:
            fields[key] = OGRTypes[type(value)]
        else:
            # Data type not supported, default to string (char 80)
            fields[key] = ogr.OFTString
        # Create the new field
        newfield = ogr.FieldDefn(key, fields[key])
        layer.CreateField(newfield)

    drv = ogr.GetDriverByName("ESRI Shapefile")
    shpdir = drv.CreateDataSource(outdir)
    # delete pre-existing output first otherwise ogr chokes
    try:
        shpdir.DeleteLayer("nodes")
    except:
        pass
    nodes = shpdir.CreateLayer("nodes", None, ogr.wkbPoint)

    # Storage for node field names and their data types
    node_fields = {}

    def create_attributes(data, fields, layer):
        attributes = {}  # storage for attribute data (indexed by field names)
        for key, value in data.items():
            # Reject spatial data not required for attribute table
            if key != "Json" and key != "Wkt" and key != "Wkb" and key != "ShpName":
                # Check/add field and data type to fields dict
                if key not in fields:
                    add_fields_to_layer(key, value, fields, layer)
                # Store the data from new field to dict for CreateLayer()
                attributes[key] = value
        return attributes, layer

    for n in G:
        data = G.nodes[n]
        g = netgeometry(n, data)
        attributes, nodes = create_attributes(data, node_fields, nodes)
        create_feature(g, nodes, attributes)

    try:
        shpdir.DeleteLayer("edges")
    except:
        pass
    edges = shpdir.CreateLayer("edges", None, ogr.wkbLineString)

    # New edge attribute write support merged into edge loop
    edge_fields = {}  # storage for field names and their data types

    for edge in G.edges(data=True):
        data = G.get_edge_data(*edge)
        g = netgeometry(edge, data)
        attributes, edges = create_attributes(edge[2], edge_fields, edges)
        create_feature(g, edges, attributes)

    nodes, edges = None, None


def info(G, n=None):
    """Return a summary of information for the graph G or a single node n.

    The summary includes the number of nodes and edges, or neighbours for a single
    node.

    Parameters
    ----------
    G : Networkx graph
       A graph
    n : node (any hashable)
       A node in the graph G

    Returns
    -------
    info : str
        A string containing the short summary

    Raises
    ------
    NetworkXError
        If n is not in the graph G

    .. deprecated:: 2.7
       ``info`` is deprecated and will be removed in NetworkX 3.0.
    """
    if n is None:
        return str(G)
    if n not in G:
        raise nx.NetworkXError(f"node {n} not in graph")
    info = ""  # append this all to a string
    info += f"Node {n} has the following properties:\n"
    info += f"Degree: {G.degree(n)}\n"
    info += "Neighbors: "
    info += " ".join(str(nbr) for nbr in G.neighbors(n))
    return info


# // END-NOSCAN
