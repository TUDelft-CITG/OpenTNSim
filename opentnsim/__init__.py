# -*- coding: utf-8 -*-

import opentnsim.core as core # now contains mixin class consumes energy, also the modified vessel properties
import opentnsim.graph_module as graph_module
import opentnsim.plot as plot
import opentnsim.model as model
import opentnsim.energy as energy
import opentnsim.strategy as strategy

"""Top-level package for OpenTNSim."""

__author__ = """Mark van Koningsveld"""
__email__ = "M.vanKoningsveld@tudelft.nl"
__version__ = "1.1.2"
