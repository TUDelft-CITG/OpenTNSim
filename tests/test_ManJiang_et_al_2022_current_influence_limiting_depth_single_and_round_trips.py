"""Here we test the estimation of fuel use with and without current influence for single and round trips in section 1 and 3."""


# Importing libraries

# package(s) related to time, space and id
import datetime, time

import pathlib

# you need these dependencies (you can get these from anaconda)
# package(s) related to the simulation
import simpy
import pandas as pd

# spatial libraries
import shapely.geometry

# package(s) for data handling
import numpy as np

# OpenTNSim
import opentnsim

# Used for making the graph to visualize our problem
import networkx as nx

import pytest

import utils


# if you wnat to make a new expected dataframe, update the expected numbers like this
# df.to_csv("test_ManJiang_et_al_2022_current_influence_limiting_depth_single_and_round_trips_expected.csv", index=False)
@pytest.fixture
def expected_df():
    path = pathlib.Path(__file__)
    return utils.get_expected_df(path)


# Creating the test objects


# Actual testing starts here
# - tests 3 fixed velocities to return the right P_tot
# - tests 3 fixed power to return indeed the same P_tot
# - tests 3 fixed power to return indeed the same v
# todo: current tests do work with vessel.h_squat=True ... issues still for False
def test_simulation(expected_df):
    # specify a number of coordinate along your route (coords are: lon, lat)
    coords = [[0, 0], [0.8983, 0], [1.7966, 0], [2.6949, 0]]

    # for each edge (between above coordinates) specify the depth (m)
    depths = [6, 2.5, 6]

    # check of nr of coords and nr of depths align
    assert len(coords) == len(depths) + 1, "nr of depths does not correspond to nr of coords"

    # create a graph based on coords and depths
    graph = nx.DiGraph()
    nodes = []
    path = []

    # add nodes
    Node = type("Site", (opentnsim.core.Identifiable, opentnsim.core.Locatable), {})

    for index, coord in enumerate(coords):
        data_node = {
            "name": "Node " + str(index),
            "geometry": shapely.geometry.Point(coord[0], coord[1]),
        }
        nodes.append(Node(**data_node))

    positions = {}
    for node in nodes:
        positions[node.name] = (node.geometry.x, node.geometry.y)
        graph.add_node(node.name, geometry=node.geometry)

    # add edges
    path = [[nodes[i], nodes[i + 1]] for i in range(len(nodes) - 1)]

    for index, edge in enumerate(path):
        # For the energy consumption calculation we add info to the graph. We need depth info for resistance.
        # NB: the CalculateEnergy routine expects the graph to have "Info" that contains "GeneralDepth"
        #     this may not be very generic!
        graph.add_edge(edge[0].name, edge[1].name, weight=1, Info={"GeneralDepth": depths[index]})

    # toggle to undirected and back to directed to make sure all edges are two way traffic
    graph = graph.to_undirected()
    graph = graph.to_directed()

    # Make your preferred class out of available mix-ins.
    TransportResource = type(
        "Vessel",
        (
            opentnsim.core.Identifiable,
            opentnsim.core.Movable,
            opentnsim.core.Routeable,
            opentnsim.core.VesselProperties,
            opentnsim.energy.ConsumesEnergy,
            opentnsim.core.ExtraMetadata,
        ),
        {},
    )

    # Create a dict with all important settings
    data_vessel = {
        "env": None,
        "name": None,
        "route": None,
        "geometry": None,
        "v": None,  # m/s
        "type": None,
        "B": 11.4,
        "L": 110,
        "H_e": None,
        "H_f": None,
        "T": 2.05,
        "safety_margin": 0.2,  # for tanker vessel with rocky bed the safety margin is recommended as 0.3 m
        "h_squat": True,  # if consider the ship squatting while moving, set to True, otherwise set to False
        "P_installed": 1750.0,  # kW
        "P_tot_given": None,  # kW
        "bulbous_bow": False,  # if a vessel has no bulbous_bow, set to False; otherwise set to True.
        "P_hotel_perc": 0.05,
        "P_hotel": None,  # None: calculate P_hotel from percentage
        "x": 2,
        "L_w": 3.0,
        "C_B": 0.85,
        "C_year": 1990,
    }

    vessel = TransportResource(**data_vessel)

    path = nx.dijkstra_path(graph, nodes[0].name, nodes[3].name)

    # Actual testing starts here
    def run_simulation(V_s, P_tot_given):
        # Start simpy environment
        simulation_start = datetime.datetime.now()
        env = simpy.Environment(initial_time=time.mktime(simulation_start.timetuple()))
        env.epoch = time.mktime(simulation_start.timetuple())

        # Add graph to environment
        env.graph = graph

        # Add environment and path to the vessel
        # create a fresh instance of vessel
        vessel = TransportResource(**data_vessel)
        vessel.env = env  # the created environment
        vessel.name = "Vessel No.1"
        vessel.route = path  # the route (the sequence of nodes, as stored as the second column in the path)
        vessel.geometry = env.graph.nodes[path[0]][
            "geometry"
        ]  # a shapely.geometry.Point(lon,lat) (here taken as the starting node of the vessel)
        vessel.v = V_s
        vessel.P_tot_given = P_tot_given

        # Start the simulation
        env.process(vessel.move())
        env.run()

        return vessel

    # prepare input data to loop through
    input_data = {"V_s": [3.5], "P_tot_given": [None]}

    # loop through the various input data
    for index, value in enumerate(input_data["V_s"]):
        # Run a basic simulation with V_s and P_tot_given combi
        vessel = run_simulation(input_data["V_s"][index], input_data["P_tot_given"][index])

        # create an EnergyCalculation object and perform energy consumption calculation
        energycalculation = opentnsim.energy.EnergyCalculation(graph, vessel)
        energycalculation.calculate_energy_consumption()

    # create dataframe from energy calculation computation
    df = pd.DataFrame.from_dict(energycalculation.energy_use)

    # add current influence
    # with current speed =0.5m/s
    U_c = 0.5
    # delta_t will be longer when upstream, shorter when downstream
    delta_t_up = df["distance"] / (df["distance"] / df["delta_t"] - U_c)
    delta_t_down = df["distance"] / (df["distance"] / df["delta_t"] + U_c)
    # total emission&fuel consumption will be larger when upstream(because of longer delta_t), smaller when downstream(because of shorter delta_t)
    df["total_fuel_consumption_kg"] = df["total_diesel_consumption_C_year_ICE_mass"] / 1000  # kg without current
    df["total_fuel_consumption_up_kg"] = df["total_diesel_consumption_C_year_ICE_mass"] / 1000 * (delta_t_up / df["delta_t"])  # kg
    df["total_fuel_consumption_down_kg"] = (
        df["total_diesel_consumption_C_year_ICE_mass"] / 1000 * (delta_t_down / df["delta_t"])
    )  # kg
    df["total_fuel_consumption_round_no_current_kg"] = df["total_diesel_consumption_C_year_ICE_mass"] / 1000 * 2  # kg
    df["total_fuel_consumption_round_current_kg"] = df["total_fuel_consumption_up_kg"] + df["total_fuel_consumption_down_kg"]  # kg

    # test the estimation of fuel consumption with and without current influence for section 1

    # if you wnat to make a new expected dataframe, update the expected numbers like this
    # df.to_csv("test_ManJiang_et_al_2022_current_influence_limiting_depth_single_and_round_trips_expected.csv", index=False)

    columns_to_test = [column for column in df.columns if "fuel" in column]
    pd.testing.assert_frame_equal(expected_df[columns_to_test], df[columns_to_test], check_exact=False)
