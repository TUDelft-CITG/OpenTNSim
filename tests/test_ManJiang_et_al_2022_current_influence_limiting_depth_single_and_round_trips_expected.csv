time_start,time_stop,edge_start,edge_stop,P_tot,P_given,P_installed,total_energy,total_diesel_consumption_C_year_ICE_mass,total_diesel_consumption_ICE_mass,total_diesel_consumption_ICE_vol,total_LH2_consumption_PEMFC_mass,total_LH2_consumption_SOFC_mass,total_LH2_consumption_PEMFC_vol,total_LH2_consumption_SOFC_vol,total_eLNG_consumption_PEMFC_mass,total_eLNG_consumption_SOFC_mass,total_eLNG_consumption_PEMFC_vol,total_eLNG_consumption_SOFC_vol,total_eLNG_consumption_ICE_mass,total_eLNG_consumption_ICE_vol,total_eMethanol_consumption_PEMFC_mass,total_eMethanol_consumption_SOFC_mass,total_eMethanol_consumption_PEMFC_vol,total_eMethanol_consumption_SOFC_vol,total_eMethanol_consumption_ICE_mass,total_eMethanol_consumption_ICE_vol,total_eNH3_consumption_PEMFC_mass,total_eNH3_consumption_SOFC_mass,total_eNH3_consumption_PEMFC_vol,total_eNH3_consumption_SOFC_vol,total_eNH3_consumption_ICE_mass,total_eNH3_consumption_ICE_vol,total_Li_NMC_Battery_mass,total_Li_NMC_Battery_vol,total_Battery2000kWh_consumption_num,total_emission_CO2,total_emission_PM10,total_emission_NOX,stationary,water depth,distance,delta_t,total_fuel_consumption_kg,total_fuel_consumption_up_kg,total_fuel_consumption_down_kg,total_fuel_consumption_round_no_current_kg,total_fuel_consumption_round_current_kg
2022-08-25 16:23:08.000000,2022-08-26 00:19:18.942451,POINT (0 0),POINT (0.8983 0),375.91458072170445,375.91458072170445,1750.0,2983.3982923032254,750912.0025832085,769683.0323768143,0.9032784581493786,199092.3117986804,211037.85050660121,2.593808287518019,2.7494367847691,498479.24683428986,528388.0016443472,1.1757002984387404,1.2462423163450649,675353.4577321368,1.5928712516115304,1212024.494130906,1284745.9637787605,1.5300655395559786,1.6218694719293374,1642084.275655836,2.0729750721988047,1297411.7383358232,1375256.4426359728,2.112065620546689,2.2387895577794907,1757769.2735493972,2.861484863917624,33902253.32162756,26.829121333662098,1.864623932689516,2382438.9900139975,1395.3572620576515,32788.58747992497,2983.3982923032254,5.8813261081800645,99998.29857959764,28570.942451,750.9120025832085,876.0640030121407,657.048002261209,1501.824005166417,1533.1120052733497
2022-08-26 00:19:18.942451,2022-08-26 08:15:29.884902,POINT (0.8983 0),POINT (1.7966 0),739.10489401719,739.10489401719,1750.0,5865.8120533937745,1369277.5217327806,1403506.2317117674,1.6471156198789543,418846.7732486713,407103.40577440953,5.456806552887619,5.303811976638433,1048691.5450511843,1019288.9783675062,2.473416837946578,2.404068702303216,1231497.5732388212,2.9045784224288567,2549835.0181317646,2478344.3166888175,3.2189239670391765,3.128673762355835,2994317.682646495,3.7800410164034903,2729471.1446537683,2652943.916298989,4.443325119203808,4.318745910254169,3205267.6563750147,5.217877580145373,66656955.15220197,52.7501083938289,3.6661325333711092,4344344.137134004,2395.742501205301,60164.82345605187,5865.8120533937745,2.2588283299637957,99998.29857959764,28570.942451,1369.2775217327805,1597.490442018655,1198.117831517827,2738.555043465561,2795.6082735364816
2022-08-26 08:15:29.884902,2022-08-26 16:11:40.827354,POINT (1.7966 0),POINT (2.6949 0),375.9145806890952,375.9145806890952,1750.0,2983.398292148847,750912.002549244,769683.0323420006,0.9032784581085226,199092.3117883782,211037.85049568088,2.5938082873838,2.749436784626828,498479.2468084956,528388.0016170053,1.1757002983779026,1.246242316280577,675353.45770159,1.5928712515394836,1212024.4940681888,1284745.9637122802,1.5300655394768041,1.6218694718454123,1642084.2755815627,2.0729750721050424,1297411.7382686876,1375256.4425648088,2.112065620437398,2.2387895576636425,1757769.2734698919,2.8614848637881964,33902253.31987326,26.8291213322738,1.8646239325930294,2382438.989906238,1395.3572620165792,32788.58747867748,2983.398292148847,5.881326108188704,99998.29857959767,28570.942452,750.912002549244,876.0640029776258,657.0480022286155,1501.824005098488,1533.1120052062413
